src/NHibernate/Hql/Ast/ANTLR/Hql.g

```diff
@@ -627,9 +627,14 @@ aggregateArgument
 	;
 
 aggregateDistinctAll
-	: ( ( DISTINCT | ALL )? ( path | collectionExpr ) )
+	: ( distinctAll aggregateArgument ) => (distinctAll aggregateArgument)
+	| aggregateArgument
```

Comment:
An obvious `( ( DISTINCT | ALL )? aggregateArgument  )` gives:
```
[fatal] rule aggregate has non-LL(*) decision due to recursive rule invocations reachable from alts 1,2.  Resolve by left-factoring or using syntactic predicates or using backtrack=true option.
```

So had to use this syntactic predicate rule with code duplication..